### ISSUE TITLE
Couldn’t connect with atlas database

### ISSUE THREAD

**Issue by Eyoatam** [denodrivers/deno_mongo#172]
when I tried to connect to an atlas database using:

```ts 
import { MongoClient } from "https://deno.land/x/mongo@v0.21.0/mod.ts";
const client = new MongoClient();
await client.connect(
	"mongodb+srv://Eyoatam:" +
		Deno.env.get("MONGO_ATLAS_PASSWORD") +
		"@cluster0.xrb4d.mongodb.net/<dbname>?retryWrites=true&w=majority"
);
```

It threw the following error: 
```bash
error: Uncaught (in promise) Error: failed to lookup address information: nodename nor servname provided, or not known
    at processResponse (deno:core/core.js:223:11)
    at Object.jsonOpAsync (deno:core/core.js:240:12)
    at async Object.connect (deno:cli/rt/30_net.js:224:13)
    at async MongoClient.connect (client.ts:41:14)
    at async app.ts:4:1
```
is there another way to connect with mongodb atlas?

**Comment by manojap** [comment]
you miss the db name

**Comment by Eyoatam** [comment]
> you miss the db name

No I didn’t , the db name is cluster0

**Comment by manojap** [comment]
do you set mongo ip ?

**Comment by manojap** [comment]
Uncaught (in promise) ConnectionRefused: No connection could be made because the target machine actively refused it. I have this issue ... while I am trying.

Any idea ? @Eyoatam

**Comment by Eyoatam** [comment]
> do you set mongo ip ?

Everything is set, I guess the problem is the library (deno_mongo), doesn’t support mongodb atlas.

**Comment by manojap** [comment]
denodb work perfect with mongo, even though I have to figure out the find ,delete operations.  The select all work fine. ( may be my fault)

**Comment by manojap** [comment]
I will post the repo when I figure it out ,

**Comment by marcodali** [comment]
Even without the "+srv" in the connection string the driver does work.
In the image attached you can see the full uri to connect without SRV displayed by Atlas to connect to replica set.

`mongodb://<username>:<password>@princesapeach-shard-00-00.h8z8n.mongodb.net:27017,princesapeach-shard-00-01.h8z8n.mongodb.net:27017,princesapeach-shard-00-02.h8z8n.mongodb.net:27017/<dbname>?ssl=true&replicaSet=PrincesaPeach-shard-0&authSource=admin&retryWrites=true&w=majority`

<img width="754" alt="Screen Shot 2021-01-24 at 22 03 47" src="https://user-images.githubusercontent.com/70110654/105660764-62747400-5e91-11eb-837a-3139eef26874.png">

```
import { MongoClient } from "https://deno.land/x/mongo@v0.21.2/mod.ts";;
import env from "./src/setup/env.ts";

const client = new MongoClient();
const url = `mongodb://${env.dbUser}:${env.dbPass}@princesapeach-shard-00-00.h8z8n.mongodb.net:27017,princesapeach-shard-00-01.h8z8n.mongodb.net:27017,princesapeach-shard-00-02.h8z8n.mongodb.net:27017/${env.dbName}?ssl=true&replicaSet=PrincesaPeach-shard-0&authSource=admin&retryWrites=true&w=majority`;
console.log('url', url);
await client.connect(url);
const names = await client.listDatabases();
console.log('names', names);
client.close();
```

```
error: Uncaught (in promise) AssertionError
    throw new AssertionError(msg);
          ^
    at assert (asserts.ts:170:11)
    at WireProtocol.receive (protocol.ts:80:7)
```

Deno 1.5.4

**Comment by manyuanrong** [comment]
@marcodali 
On the basis that you get the complete node information. Currently, only one node can be selected to connect (please select the master node), 
for example:
```ts
const client = new MongoClient();
const url = `mongodb://user:pwd@princesapeach-shard-00-02.h8z8n.mongodb.net:27017/db?retryWrites=true&w=majority`;
await client.connect(url);
```

**Comment by berkaygure** [comment]
Hi everyone, I was trying to connect to the Atlas database. I have tried all the possible solutions above. finally, I can connect with this way
```javascript
await client.connect({
    db:'<db>',
    tls:true,
    servers: [
        {
            host: 'cluster0-shard-00-02.soz4m.mongodb.net',
            port: 27017
        }
    ],
    credential: {
        username:"<username>",
        password: '<pwd>',
        db: '<db>',
        mechanism:'SCRAM-SHA-1'
    }
})

```

**Comment by dkaerit** [comment]
This don't work for me.

They can't somehow fix it so that it gets overloaded and the uri can be put back as a string, like it could be done before

**Comment by FSou1** [comment]
> Hi everyone, I was trying to connect to the Atlas database. I have tried all the possible solutions above. finally, I can connect with this way
> 
> ```js
> await client.connect({
>     db:'<db>',
>     tls:true,
>     servers: [
>         {
>             host: 'cluster0-shard-00-02.soz4m.mongodb.net',
>             port: 27017
>         }
>     ],
>     credential: {
>         username:"<username>",
>         password: '<pwd>',
>         db: '<db>',
>         mechanism:'SCRAM-SHA-1'
>     }
> })
> ```

I haven't had any success.

It is either:
```
Error: MongoError: {"topologyVersion":{"processId":"601480457f8872e51ff2e12f","counter":4}
```
or
```
Error: MongoError: {"ok":0,"errmsg":"noTimeout cursors are disallowed in this atlas tier","code":8000,"codeName":"AtlasError"}
```

**Comment by berkaygure** [comment]
You should set your primary cluster address to host. After that, It should connect successfully. I think you are getting an error while finding objects. 

Can you try this? It works for me.

```typescript
interface Room {
  _id?: { $oid: string };
  name: string;
  owner_id: string;
}

const rooms = db.collection<Room>('rooms');
const myRoom = rooms.find({}, {
      noCursorTimeout: false,
} as any);

```

**Comment by FSou1** [comment]
@berkaygure Yeah, it works, thank you.

@manyuanrong Any reason not to expose the `noCursorTimeout` property of the `FindOptions` interface?

**Comment by ShubhamPalriwala** [comment]
If @berkaygure approach still gives the same `Error: failed to lookup address information: nodename nor servname provided, or not known`
Make sure your host URL is taken from the below steps:

 ![hm](https://user-images.githubusercontent.com/53137941/107846535-c9c47a80-6de4-11eb-8a32-6a87de81fd75.png)

Credits @tavadur

**Comment by nooobcoder** [comment]
Just tried connecting again with deno_mongo
Did not work out

ATTACHING RELEVANT CODE AND SCREENSHOT BELOW
`import { MongoClient } from "../deps.ts";
import "https://deno.land/x/dotenv/load.ts";

const client = new MongoClient();
client.connect(
	"mongodb://admin:adminadmin@merncluster-shared-bla-01.blablabla.mongodb.net:27017"
);
// DB URL IS FAKE
const db = client.database("deno_survey");
export const usersCollection = db.collection("users");
`

**Comment by nooobcoder** [comment]
@manojap @Eyoatam @ShubhamPalriwala @manyuanrong can  you guys help me please?

**Comment by ShubhamPalriwala** [comment]
> Hi everyone, I was trying to connect to the Atlas database. I have tried all the possible solutions above. finally, I can connect with this way
> 
> ```js
> await client.connect({
>     db:'<db>',
>     tls:true,
>     servers: [
>         {
>             host: 'cluster0-shard-00-02.soz4m.mongodb.net',
>             port: 27017
>         }
>     ],
>     credential: {
>         username:"<username>",
>         password: '<pwd>',
>         db: '<db>',
>         mechanism:'SCRAM-SHA-1'
>     }
> })
> ```

Try this @nooobcoder 
Credits @berkaygure

**Comment by nooobcoder** [comment]
> > Hi everyone, I was trying to connect to the Atlas database. I have tried all the possible solutions above. finally, I can connect with this way
> > ```js
> > await client.connect({
> >     db:'<db>',
> >     tls:true,
> >     servers: [
> >         {
> >             host: 'cluster0-shard-00-02.soz4m.mongodb.net',
> >             port: 27017
> >         }
> >     ],
> >     credential: {
> >         username:"<username>",
> >         password: '<pwd>',
> >         db: '<db>',
> >         mechanism:'SCRAM-SHA-1'
> >     }
> > })
> > ```
> 
> Try this @nooobcoder
> Credits @berkaygure

@ShubhamPalriwala what should i write here? What to write in place of `<db>` ?
`db:'<db>',`

**Comment by ShubhamPalriwala** [comment]
@nooobcoder write your DB name that you initialised on Mongo Atlas

**Comment by dkaerit** [comment]
Someone already has a functional way to connect to mongo atlas with this module, it is a bit stressful to want to connect mongo atlas and not have a module in deno that works correctly

**Comment by manyuanrong** [comment]
closed by #198

**Comment by juratori** [comment]
> Someone already has a functional way to connect to mongo atlas with this module, it is a bit stressful to want to connect mongo atlas and not have a module in deno that works correctly

Although pr #198 was meant to resolve the MongoDB Atlas connection problem, it did not work as expected in my case. As a workaround, I have implemented the following code that effectively determines the active and operational replica among the three replica instances in MongoDB Atlas shared clusters.

```
/**
 * Determines the active and functioning replica among the three
 * replica instances in MongoDB Atlas shared clusters.
 * The replica instances take turns being active at different
 * times as they rotate. The script establishes a connection with
 * each replica instance and checks whether it is available and
 * responsive. If an active replica is found, its URI is returned.
 * If none of the replica instances are available, an error is thrown.
 *
 * @param mongoUri - The URI of the replica instance to check.
 * @returns The URI of the active replica instance.
 */

import { MongoClient } from "../../deps.ts";

import {
  mongoUri1,
  mongoUri2,
  mongoUri3,
  mongoDbName,
  mongoDbUsername,
  mongoDbPassword,
} from "../config/global.ts";

const mongoUris = [mongoUri1, mongoUri2, mongoUri3];

async function checkEndpoint(mongoUri: string) {
  const client = new MongoClient();
  try {
    await client.connect({
      db: mongoDbName,
      tls: true,
      servers: [
        {
          host: mongoUri,
          port: 27017,
        },
      ],
      credential: {
        username: mongoDbUsername,
        password: mongoDbPassword,
        db: mongoDbName,
        mechanism: "SCRAM-SHA-1",
      },
    });
    await client.listDatabases();
    return mongoUri;
  } catch {
    return null;
  } finally {
    await client.close();
  }
}

async function checkEndpoints() {
  const endpointChecks = mongoUris.map((mongoUri) => checkEndpoint(mongoUri));
  const results = await Promise.allSettled(endpointChecks);
  const availableUri = results
    .filter((result) => result.status === "fulfilled")
    .map((result) => (result as PromiseFulfilledResult<string | null>).value)
    .find(Boolean);

  if (availableUri) {
    return availableUri;
  }

  throw new Error("No mongoUri is available");
}

export default checkEndpoints;
```
